### ISSUE TITLE
Spacing with JSX props

### ISSUE THREAD

**Issue by oyeanuj** [prettier/prettier#3062]
**Prettier 1.7.4**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22function%20HelloWorld(%7Bgreeting%20%3D%20%5C%22hello%5C%22%2C%20greeted%20%3D%20'%5C%22World%5C%22'%7D)%20%7B%5Cn%5Cn%20%20return%20(%5Cn%20%20%5Ct%3Cdiv%20%5Cn%20%20%20%20%20%20className%5Ct%5Ct%3D%20'HelloWorld'%20%5Cn%20%20%20%20%20%20title%5Ct%5Ct%5Ct%3D%20'How%20are%20you%3F'%5Cn%20%20%20%20%20%20onMouseOver%5Ct%3D%20%7BonMouseOver%7D%5Cn%20%20%20%20%3E%5Cn%5Ct%5CtHello%20world%5Cn%20%20%20%20%3C%2Fdiv%3E%5Cn%20%20)%3B%5Cn%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Afalse%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Afalse%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A4%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Atrue%7D%7D)
```sh
--no-bracket-spacing
--no-semi
--tab-width 4
--use-tabs
```

**Input:**
```jsx
function HelloWorld({greeting = "hello", greeted = '"World"'}) {

  return (
    <div 
      className		= 'HelloWorld' 
      title		= 'How are you?'
      onMouseOver	= {onMouseOver}
    >
      Hello world
    </div>
  );

}
```

**Output:**
```jsx
function HelloWorld({greeting = "hello", greeted = '"World"'}) {
  return (
    <div
	className="HelloWorld"
	title="How are you?"
	onMouseOver={onMouseOver}
    >
	Hello world
    </div>
  )
}

```

*Apologies if this issue has already been raised, searched but couldn't find anything similar*.

Thank you for this excellent project. While integrating in our projects, found ourselves hitting a formatting issue.

So, in our projects, we write JSX like shown above (Github formatting messes it a little, so the playground link is more appropriate) . While our style preferences use tabs, the idea is that the props are easily scannable, since the eye doesn't have to zigzag. 

On using Prettier, the JSX becomes like shown above. It does two things in particular - strips out tabs and spacing in between prop and prop values, as well as around the `=` sign. The second in inconsistent with how it treats function parameters as shown above.

So, my question would be:

1. Is there a way to configure Prettier to always produce output like the given input (aligned)?
2. If not, is there or can there be a way to tell Prettier to not mess with JSX spacing?
3. Why does the `=` not have a space before and after in props, like it does in function parameters?

Thanks again!

**Comment by lydell** [comment]
> Is there a way to configure Prettier to always produce output like the given input (aligned)?

No. Prettier does not do alignment, because changing a single line can cause several other lines to be re-formatted, causing unnecessary git diff. See for example #188, #350 and #1622.

> If not, is there or can there be a way to tell Prettier to not mess with JSX spacing?

No. Well, you _could_ add `// prettier-ignore` comments all over the place ;)

> Why does the `=` not have a space before and after in props, like it does in function parameters?

Because that's how the majority of all HTML is written like that.  People carried over that style to JSX. Prettier has chosen the most common JSX style.

**Comment by oyeanuj** [comment]
@lydell Thanks for the quick response! Good to know that I am not the only one who prefers vertical alignment. One quick follow up though:

> Because that's how the majority of all HTML is written like that. People carried over that style to JSX. Prettier has chosen the most common JSX style.

Is this something that can be considered as a feature request, to either ignore, or to allow for spaces given that it already does it for JS function parameters?

**Comment by lydell** [comment]
@oyeanuj I've never heard about anyone putting spaces around `=` in HTML or JSX, so I doubt that such a feature would ever get enough support to warrant adding it.

I used to be an adamant "no semicolons" person. Then I started using Prettier (before Prettier had `--no-semi`). It annoyed me seeing semicolons everywhere in the beginning, but I always reminded myself that they were a cheap price to pay for all the other things I liked about Prettier. These days I've grown accustomed to semicolons and actually prefer them.

I'm sure the same can happen to you regarding spaces around `=`. If not, then it might be that an automatic, opinionated code formatter is not for you. Just my two cents.
